posts/crud-pattern.md

```diff
@@ -0,0 +1,294 @@
+.. title: CRUD Pattern
+.. slug: crud-pattern
+.. date: 2020-03-12 17:02:28 UTC+05:30
+.. tags:
+.. category:
+.. link:
+.. description:
+.. status: draft
+.. type: text
+
+# CRUD Pattern
+
+React.js based frontend powered by REST API can be a typical architecture in Dashboard like applications. In my last project, I was working on such a Dashboard application where front-end was written in React.js and the API was powered by Django REST framework. The application consisted of multiple pages with a similar structure. Each page would look like a table of records which can be **Created**, **Retrieved**, **Updated** or **Deleted**, the common **CRUD** pattern. When I joined the project, the prototype was already in place and the code had become complex to understand. After several sessions of work on different parts of the codebase, it was apparent that each page had similar functionality(CRUD actions). There seemed a possibility of abstraction of this functionality into a common component that would provide a neatly structured code that is easier to understand and maintain.
+
+While 'thinking' about the structure of the abstract component, a conclusion was formed that each such component required -
+
+- A Table UI that will render a short view of the retrieved object. it will have an interface to Update or Delete the object.
+- A Form UI that will show details of the object in an editable view. It will provide UI to either Create a new object or Update the existing object.
+
+In this blog post, we will go through the implementation of the CRUD pattern using a simple Library App. We'll look in detail how the functionality can be split into multiple components. We'll check the Form component that handles all of its required functionality (onChangeHandler, formValidator etc.) by itself. While the eventual implementation has undergone several iterations, we will also check a couple of false starts which we started with and refactored the components properly later on.
+
+We'll structure our pattern in the following way -
```

Comment:
This is directly getting into too many details. It should flow like an overview - and then details.